### ISSUE TITLE
Grammar-guided generation support

### ISSUE THREAD

**Issue by tmc** [jmorganca/ollama#808]
Thoughts on introducing a straightforward way for a Modelfile to point to a grammar and thread that through to sampling/inference?

**Comment by Lwrless** [comment]
I'd like that feature too.
I see that grammar support was added in PR https://github.com/jmorganca/ollama/pull/565, but It has not been merged yet.

**Comment by briancleland** [comment]
Another vote for this feature to be merged.

**Comment by igorschlum** [comment]
I asked bing for more explanations and I agree.

Okay, I will try to translate what I just said in English, including the description of a Modelfile. Here is what I get:

A Modelfile is a file that contains the parameters of a machine learning model, as I explained to you earlier. A machine learning model is an algorithm that learns from data and that can perform tasks such as classification, regression, generation, etc. A Modelfile allows you to save and load a machine learning model, in order to reuse it or share it .

A grammar is a set of rules that define the structure and meaning of a language, such as French or English. Sampling or sampling is a method that allows you to generate text from a machine learning model, by randomly choosing the words or characters to add to the text. Inference is a method that allows you to generate text from a machine learning model, by choosing the words or characters that maximize the probability of the text.

What this user would like is that there is a simple way for a Modelfile to indicate which grammar it refers to, and that this grammar is taken into account during sampling or inference. Thus, the generated text would be more coherent and more compliant with the rules of the language.

For example, if the Modelfile corresponds to a model that generates French, it could point to a grammar that specifies the rules of conjugation, agreement, punctuation, etc. of French. During sampling or inference, the model would use this grammar to generate correct text in French.

**Comment by zabirauf** [comment]
+1 to adding this in ollama

**Comment by donedgardo** [comment]
+1 would love to see this added to support memgpt integration

**Comment by orkutmuratyilmaz** [comment]
+1 to this feature with MemGPT integration:)

**Comment by adambuttrick** [comment]
+1 to this feature for my use cases. I get the odd, incorrectly formatted response using JSON mode and the output format in the prompt, so I think this would greatly improve reliability.

**Comment by tezlm** [comment]
It would also be nice to be able to change the grammar on the fly, instead of baked into a Modelfile

**Comment by technovangelist** [comment]
The original issue asked for grammar guided generation. This has been added recently with `format: json` available at the API and the CLI. You can output web formed JSON and specify the schema to be used for that output. So I will go ahead and close the issue now. If you think there is anything we left out, reopen and we can address. Thanks for being part of this great community.

**Comment by hobofan** [comment]
@technovangelist I think this issue should be reopened. `format: json` only allows for specifying structure for JSON output, but does not give access to the underlying `grammar` parameter of llama.cpp to allow for doing grammar guided generation according to different GBNF grammars (like implemented in this PR: https://github.com/jmorganca/ollama/pull/565).

E.g. one might want to use a custom GBNF grammer to generate answers in different data formats (e.g. YAML or something more esoteric).

**Comment by technovangelist** [comment]
ok, that makes sense. Since the original issue wasn't about GBNF specifically and your comment is, would you mind opening a new issue making the case for a lower level GBNF feature? I could make it, but then it may be harder for you to follow up on it. 

Thanks so much for clarifying and adding this info.

**Comment by James4Ever0** [comment]
Grammar support would be awesome, and logits probability response could be a good starting point.

**Comment by ggregoire** [comment]
> The original issue asked for grammar guided generation. This has been added recently with format: json available at the API and the CLI. **You can output web formed JSON and specify the schema to be used for that output.** So I will go ahead and close the issue now. If you think there is anything we left out, reopen and we can address. Thanks for being part of this great community.

Hi @technovangelist ! Maybe I missed it in the docs, but how do you specify the schema? I can't find the option in https://github.com/ollama/ollama/blob/main/docs/api.md#generate-a-completion. Also, it seems like the pull request that adds support for the JSON schema has never been merged: https://github.com/ollama/ollama/pull/830 (maybe another one was merged instead?). Thanks!
